testsuite/tests/comprehensions/array_comprehensions_pure.ml

```diff
@@ -0,0 +1,378 @@
+|}];;
+
+[|i,s for i = 0 to 3 for s in [|'a'; 'b'; 'c'|]|];;
+[%%expect{|
+- : (int * char) array =
+[|(0, 'a'); (0, 'b'); (0, 'c'); (1, 'a'); (1, 'b'); (1, 'c'); (2, 'a');
+  (2, 'b'); (2, 'c'); (3, 'a'); (3, 'b'); (3, 'c')|]
+|}];;
+
+[|i,j for i = 1 to 3 for j in [|i*10; i*100|]|];;
+[%%expect{|
+- : (int * int) array =
+[|(1, 10); (1, 100); (2, 20); (2, 200); (3, 30); (3, 300)|]
+|}];;
+
+[|x for xs in [|[|"this"; "is"; "one"|];
+                [|"way"|];
+                [|"to"; "flatten"|];
+                [|"a"; "nested"; "array"|]|]
+    for x in xs|];;
+[%%expect{|
+- : string array =
+[|"this"; "is"; "one"; "way"; "to"; "flatten"; "a"; "nested"; "array"|]
+|}];;
+
+[|i for i = 0 to 10 when i mod 2 = 0|];;
+[%%expect{|
+- : int array = [|0; 2; 4; 6; 8; 10|]
+|}];;
+
+[|() for _ = 1 to 10|];;
+[%%expect{|
+- : unit array = [|(); (); (); (); (); (); (); (); (); ()|]
+|}];;
+
+(******************************************************************************)
+(**** More complex behavior ****)
+
+let pythagorean_triples n =
+  [|a,b,c for a = 1 to n for b = a to n for c = b to n when a*a + b*b = c*c|]
+in
+pythagorean_triples 10;;
+[%%expect{|
+- : (int * int * int) array = [|(3, 4, 5); (6, 8, 10)|]
+|}];;
+
+let tails xs =
+  let len = Array.length xs in
+  Array.init (len + 1) (fun i -> Array.sub xs i (len - i))
+in
+let sum = Array.fold_left ( + ) 0 in
+[|sum xs for xs in tails [|1; 20; 300; 4_000; 50_000; 600_000; 7_000_000|]|];;
+[%%expect{|
+- : int array =
+[|7654321; 7654320; 7654300; 7654000; 7650000; 7600000; 7000000; 0|]
+|}];;
+
+let xs = [|2;7;18;28|] in
+[|x + 1000*y for x in xs and y in xs|];;
+[%%expect{|
+- : int array =
+[|2002; 7002; 18002; 28002; 2007; 7007; 18007; 28007; 2018; 7018; 18018;
+  28018; 2028; 7028; 18028; 28028|]
+|}];;
+
+(******************************************************************************)
+(**** Edge cases ****)
+
+[|42 when true|];;
+[%%expect{|
+- : int array = [|42|]
+|}];;
+
+[|42 when false|];;
+[%%expect{|
+- : int array = [||]
+|}];;
+
+[|x for x in [||]|];;
+[%%expect{|
+- : '_weak1 array = [||]
+|}];;
+
+[|i for i = 0 to -1|];;
+[%%expect{|
+- : int array = [||]
+|}];;
+
+[|i for i = 0 downto 1|];;
+[%%expect{|
+- : int array = [||]
+|}];;
+
+[|i for i = 0 to 0|];;
+[%%expect{|
+- : int array = [|0|]
+|}];;
+
+[|i for i = 0 downto 0|];;
+[%%expect{|
+- : int array = [|0|]
+|}];;
+
+(* This would take ~forever if the empty array were iterated over later;
+   however, for arrays, using [and] lets us get more flexibility (see below). *)
+[|i,j,k for i in [||] for j = 0 to Int.max_int for k = 0 downto Int.min_int|];;
+[%%expect{|
+- : ('_weak2 * int * int) array = [||]
+|}];;
+
+(* This works no matter where the empty array is, but would take ~forever for
+   lists if the empty list were iterated over later; see
+   "array_comprehensions_special.ml" for more nuance on what can happen here
+   with arrays. *)
+[|i,j,k for i in [||] and j = 0 to Int.max_int and k = 0 downto Int.min_int|];;
```

Comment:
That goes in `array_comprehensions_special.ml`, since the equivalent of this with an empty list towards the front would effectively hang for lists and the non-special array tests are mirrored with the list tests.